Migration guide for Service ExternalName Istio changed behavior (#975)
kyma-project/istio: docs/user/migration/external-name-alias.md

Author: let's proceed as we discussed

Author: let's proceed as was discussed

Author: let's proceed as was discussed

# Migration Guide for External Name Alias's Behavior Change

## Context

Istio with 1.21 version has changed the behavior of the Service of type ExternalName.

> Previously, for the Service of type ExternalName a separate `cluster` in the Envoy config was created.

Reviewer: ```suggestion
```

Author: Why do we want to remove that?

Author: Especially line bellow we have a continuation, so removing this one would make text broken

Author: I think it's important to mention that in the context

Reviewer: This seems kind of out of place to me, as what the `cluster` thing does is never explained. And I have doubts that Istio module users understand what that means.

Author: I think for advanced user it might be beneficial to understand better the root of the change. For less advanced users, it does not hurt possibility to understand the rest, and can be easily skipped as a detail.

With the change, the Service of type ExternalName is treated as an alias of the Service that it points to.

> This change has been introduced to align the behavior of Istio with handling of the Service of type ExternalName with the Kubernetes behavior.

Reviewer 2: ```suggestion
Istio introduced this change to align the way it handles the Service of type ExternalName with the Kubernetes behavior.
```

> Since it caused some issues for our users, we introduced a new annotation "disable-external-name-alias" to disable this behavior.

Reviewer 2: ```suggestion
Since it caused some issues for our users, we introduced a new annotation, **disable-external-name-alias**, to disable this behavior.
```

> Since it caused some issues for our users, we introduced a new annotation, **disable-external-name-alias**, to disable this behavior.
> If you are affected by this change, update your configuration to align with the new behavior.

Reviewer: ```suggestion
Since it caused some issues for our users, we introduced a new annotation, **disable-external-name-alias**, to disable this change.
However, use of this parameter is not recommended, and Istio provides a migration path from the previous behaviour.
If you are using this parameter, please update your configuration according to the following points.
```

Author: This way, I think we are encountering a problem that the Istio migration guide from 1.21 upgrade notes does not fully work. I think mentioning that might be harmful. Someone might go there and use it.

> If you are affected by this change, you should update your configuration to align with the new behavior.

Reviewer 2: ```suggestion
If you are affected by this change, update your configuration to align with the new behavior.
```

## VirtualService

> According to [Istio 1.21 Upgrade Notes](https://istio.io/latest/news/releases/1.21.x/announcing-1.21/upgrade-notes/), VirtualService cannot point with its destination to the Service of type ExternalName.

Reviewer 2: ```suggestion
According to [Istio 1.21 Upgrade Notes](https://istio.io/latest/news/releases/1.21.x/announcing-1.21/upgrade-notes/), the destination of a VirtualService cannot point to the Service of type ExternalName.
```

The limitation exists in the traffic coming from the internal mesh clients.

> Creating a ServiceEntry, has no effect despite upgrade notes for version 1.21.

Reviewer: ```suggestion
Upgrade notes for version 1.21 describe a migration path by creating a ServiceEntry, although in currently released Istio versions it has no effect .
```

> Creating a ServiceEntry, has no effect despite upgrade notes for 1.21.

Reviewer 2: ```suggestion
Creating a ServiceEntry, has no effect despite upgrade notes for version 1.21.
```

> Routing to the Service of type ExternalName which alias other type of Service also will not work.

Reviewer 2: ```suggestion
Routing to the Service of type ExternalName, which aliases another type of Service, will also not work.
```

> Instead, it has to point to the actual host that is being aliased by the Service of type ExternalName.

Reviewer 2: ```suggestion
Instead, it has to point to the actual host aliased by the Service of type ExternalName.
```

The following configuration will not work:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: ext-svc
  namespace: default
spec:
  type: ExternalName
  externalName: httpbin.org

---

apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: destination-rule
  namespace: default
spec:
  host: ext-svc.default.svc.cluster.local
  trafficPolicy:
    loadBalancer:
      simple: ROUND_ROBIN
```

Use the following configuration instead:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: external-httpbin
  namespace: default
spec:
  hosts:
    - httpbin.org
  ports:
    - number: 80
      name: http
      protocol: HTTP
  resolution: DNS
  location: MESH_EXTERNAL

---

apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: destination-rule
  namespace: default
spec:
  host: httpbin.org
```

Remember to test if the new configuration works as expected and fits all your needs.

## Service ExternalName Ports

In the new behavior, Istio ignores the ports set in the Service of type ExternalName.

> In case you rely on them in any way, you must update your configuration to align with the new behavior.

Reviewer: What needs to be done to align with this behaviour?

Author: depending on the use case

Author: aligned to what we discussed

# Migration guide for external name alias behavior change

## Context

Istio with 1.21 version has changed the behavior of the Service of type ExternalName.
Previously, for the Service of type ExternalName a separate `cluster` in the Envoy config was created.
With the change, the Service of type ExternalName is treated as an alias of the Service that it points to.
This change has been introduced to align the behavior of Istio with handling of the Service of type ExternalName with the Kubernetes behavior.
Since it caused some issues for our users, we introduced a new annotation "disable-external-name-alias" to disable this behavior.
Under the hood, we apply `ENABLE_EXTERNAL_NAME_ALIAS` environment variable on the Istiod component.
At certain point Istio will drop the support for this environment variable, making the behavior change obligatory.

> Currently, it's already removed from the main branch with the [PR](https://github.com/istio/istio/pull/52317).

Reviewer: I would not mention that in this customer facing documentation, the state of the PR might change, and anyway it needs to have a release and we need to bump that

Author: This PR is already merged.

Author: We can remove this from this tutorial, I put it here to bring more context and insight into what's the situation. Unfortunately it is already dropped and I doubt if it will be back again.

We expect it to be a part of the 1.24 release.
Due to that, deprecation on the annotation will be introduced.
If you are affected by this change, you should update your configuration to align with the new behavior.

## Virtual Service

Due to the current Istio misbehavior, VirtualService cannot point with its destination to the Service of type ExternalName as the upgrade notes states [Ref](https://istio.io/latest/news/releases/1.21.x/announcing-1.21/upgrade-notes/).
The limitation exists in the traffic coming from the internal mesh clients.
Creating a ServiceEntry, has no effect despite upgrade notes for 1.21.
Routing to the Service of type ExternalName which alias other type of Service also will not work.
Instead, it has to point to the actual host that is being aliased by the Service of type ExternalName.

> With [PR](https://github.com/istio/istio/pull/52589) merged, VirtualService should be able to point to the Service of type ExternalName for the internal mesh traffic.

Reviewer: I would not mention that in documentation, the state of the PR might change, and anyway it needs to have a release and we need to bump that

Author: Yeah, that's right. I'm only concerned about the context. Certain parts of the migration guide might not be fully reflecting the state after the fix. Do you have any suggestions how'd you handle it here?

Author: Should we somehow mention, that the current steps are adjusted to the Istio not intended behavior atm?

Author: let's proceed as was discussed
